### PR TITLE
Actor upgrade workflow: the become syscall

### PR DESCRIPTION
**Pull request by vyzo**
Heavy WIP -- doesn't compile yet due to XXXs.

### CODE REVIEW

**Comment by mriise** [comment]
so the syscall itself (from wasm perspective) would look like `fn become_actor(from: (CID)) -> !` where `!` is important inside WASM to trigger the trap code in wasmtime

from the FVM side of things, it would look more like `fn become_actor(context: Context<'_, impl Kernel>, from: (CID)) -> Result<Never, Abort>`

**Comment by mriise** [comment]
@vyzo  why is the tuple important here?

**Comment by vyzo** [comment]
extensibility and API consistency with the other entry points.

File: fvm/src/syscalls/actor.rs [hunk 1]
```diff
@@ -109,3 +112,19 @@ pub fn get_code_cid_for_type(
...
+    // TODO different exit code?
+    let cid = context
+        .memory
+        .read_cid(code_cid_off)
+        .map_err(|e| Abort::from_error(ExitCode::USR_ILLEGAL_ARGUMENT, e))?;
+    context.kernel.become_actor(&cid)?; // ExitCode::SYS_INVALID_RECEIVER
+    Err(Abort::Exit(
+        ExitCode::OK,
+        "became a new actor something something".to_string(),
+    ))
```
Comment (Stebalien) [inline fvm/src/syscalls/actor.rs:129]: Hm. So, the tricky part will be returning a value.

Comment (Stebalien) [reply fvm/src/syscalls/actor.rs:129]: We probably need a new variant in `Abort::` along the lines of `Abort::Return`. (and maybe we should rename `Abort` to `EarlyExit`, or something like that...

File: sdk/src/sys/actor.rs [hunk 2]
```diff
@@ -95,4 +95,8 @@ super::fvm_syscalls! {
     /// **Privledged:** May only be called by the init actor.
     #[doc(hidden)]
     pub fn create_actor(actor_id: u64, typ_off: *const u8) -> Result<()>;
+
+    /// TODO
+    #[doc(hidden)]
+    pub fn become_actor(cid_off: *const u8) -> !;
```
Comment (Stebalien) [inline sdk/src/sys/actor.rs:101]: This is `Result<!>` because it _can_ fail.

File: fvm/src/kernel/mod.rs [hunk 3]
```diff
@@ -184,6 +185,10 @@ pub trait ActorOps {
 
     /// Returns the CodeCID for the supplied built-in actor type.
     fn get_code_cid_for_type(&self, typ: actor::builtin::Type) -> Result<Cid>;
+
+    /// Become changes the code of the running actor using the specified code cid;
+    /// it does not return unless there is an error
+    fn become_actor(&self, code_cid: &Cid) -> std::result::Result<(), Abort>;
```
Comment (Stebalien) [inline fvm/src/kernel/mod.rs:191]: This should probably return `Result<BlockId>`.

- If it succeeds, we abort with the `BlockId` as the result.
- If it fails, we return a syscall error to the user.

File: sdk/src/sys/actor.rs (hunk 2 above)
Comment (mriise) [reply sdk/src/sys/actor.rs:101]: Result<!> requires nightly feature flag

Comment (vyzo) [reply sdk/src/sys/actor.rs:101]: I have already changed this, pls stop.

Comment (mriise) [reply sdk/src/sys/actor.rs:101]: Result<Infallible> should work the same though

Comment (Stebalien) [reply sdk/src/sys/actor.rs:101]: Yep, either way.

File: fvm/src/syscalls/actor.rs (hunk 1 above)
Comment (mriise) [reply fvm/src/syscalls/actor.rs:129]: I like `Terminate`

File: fvm/src/call_manager/default.rs [hunk 4]
```diff
@@ -148,10 +150,148 @@ where
...
+                // TODO charge gas
+                // Charge for any remaining uncharged execution gas, returning an error if we run
+                // out.
+                // charge_for_exec(&mut store)?;
+
+                // If the invocation failed due to running out of exec_units, we have already
+                // detected it and returned OutOfGas above. Any other invocation failure is returned
+                // here as an Abort
+                Ok(())
+            })();
```
Comment (mriise) [inline fvm/src/call_manager/default.rs:284]: can use `let res: Result<(), Abort> = { <code> Ok(()) }` instead of closure

Comment (mriise) [reply fvm/src/call_manager/default.rs:284]: nvm, we are using `?` opeator

File: fvm/src/call_manager/default.rs [hunk 5]
```diff
@@ -430,6 +570,13 @@ where
                             "fatal error".to_owned(),
                             Err(ExecutionError::Fatal(err)),
                         ),
+                        // XXX this really is not possible dear compiler, Abort::Return
```
Comment (mriise) [inline fvm/src/call_manager/default.rs:573]: if something really isn't reachable, you can tell the compiler with `unreachable!()`
```rust 
match result {
    Ok(_) => unreachable!(),
    Err(e) => e,
}

**Pull request closed without merging**